pkg/provider/kubernetes/metering.go

```diff
@@ -0,0 +1,175 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"github.com/minio/minio-go"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/provider"
+	"k8c.io/kubermatic/v2/pkg/util/errors"
+)
+
+const AccessKey = "accessKey"
+const SecretKey = "secretKey"
+const Bucket = "bucket"
+const Endpoint = "endpoint"
+const ReportPrefix = "report-"
+const SecretName = "s3"
+const SecretNamespace = "kubermatic"
+
+var urlValidTime = time.Hour * 1
+
+var secretNamespacedName = types.NamespacedName{Name: SecretName, Namespace: SecretNamespace}
+
+// MeteringProvider struct that holds required components in order to manage metering objects.
+type MeteringProvider struct {
+	seedsGetter      provider.SeedsGetter
+	seedClientGetter provider.SeedClientGetter
+}
+
+// NewMeteringProvider returns a metering provider
+func NewMeteringProvider(seedsGetter provider.SeedsGetter, seedKubeconfigGetter provider.SeedClientGetter) *MeteringProvider {
+	return &MeteringProvider{
+		seedsGetter:      seedsGetter,
+		seedClientGetter: seedKubeconfigGetter,
+	}
+}
+
+func MeteringProviderFactory(seedsGetter provider.SeedsGetter, seedClientGetter provider.SeedClientGetter) provider.MeteringProviderGetter {
+	return func() (provider.MeteringProvider, error) {
+		return NewMeteringProvider(
+			seedsGetter,
+			seedClientGetter,
+		), nil
+	}
+}
+
+func (p *MeteringProvider) GetReport(ctx context.Context, reportName string) (*string, error) {
+
+	seedsMap, err := p.seedsGetter()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, seed := range seedsMap {
+
+		mc, bucket, err := p.getS3DataFromSeed(ctx, seed)
+		if err != nil {
+			return nil, err
+		}
+		_, err = mc.StatObject(*bucket, reportName, minio.StatObjectOptions{})
+		if err != nil {
+			continue
+		}
+
+		presignedURL, err := mc.PresignedGetObject(*bucket, reportName, urlValidTime, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		urlString := presignedURL.String()
+		return &urlString, nil
+	}
+
+	return nil, errors.New(404, "report not found")
+}
+
+// ListReports returns a list of all reports generated by metering
+func (p *MeteringProvider) ListReports(ctx context.Context) ([]string, error) {
+
+	seedsMap, err := p.seedsGetter()
+	if err != nil {
+		return nil, err
+	}
+
+	var combinedReports []string
+	for _, seed := range seedsMap {
+
+		reports, err := p.getReportsForSeed(ctx, seed)
+		if err != nil {
+			return nil, err
+		}
+		combinedReports = append(combinedReports, reports...)
+
+	}
+
+	// In-place deduplicate
+	if len(combinedReports) > 1 {
```

Comment:
This part seems a bit overcomplicated and the complexity depending on sort would be in the best scenario probably `O(n * log(n))` + `O(n)` to traverse the list. I'd simply do:

```go
result := make([]string, 0)
unique := make(map[string]struct{})

for _, report := range combinedReports {
	if _, exists := unique[report]; !exists {
		result = append(result, report)
	}

	unique[report] = struct{}{}
}

return result
```